sdk/iot/Azure.Iot.Hub.Service/src/Authentication/SasTokenProviderWithSharedAccessKey.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+
+namespace Azure.Iot.Hub.Service.Authentication
+{
+    /// <summary>
+    /// Implementation of a shared access signature provider with token caching and refresh.
+    /// </summary>
+    internal class SasTokenProviderWithSharedAccessKey : ISasTokenProvider
+    {
+        private static readonly TimeSpan s_defaultTimeToLive = TimeSpan.FromMinutes(30);
+        private readonly object _lock = new object();
+
+        private readonly string _hostname;
+        private readonly string _sharedAccessPolicy;
+        private readonly string _sharedAccessKey;
+        private readonly TimeSpan _timeToLive;
+
+        private string _cachedSasToken;
+        private DateTimeOffset _tokenExpiryTime;
+
+        // Private constructor to prevent accidental initialzation.
+        private SasTokenProviderWithSharedAccessKey()
+        {
+        }
+
+        public SasTokenProviderWithSharedAccessKey(string hostname, string sharedAccessPolicy, string sharedAccessKey, TimeSpan? timeToLive = null)
+        {
+            _hostname = hostname;
+            _sharedAccessPolicy = sharedAccessPolicy;
+            _sharedAccessKey = sharedAccessKey;
+            _timeToLive = timeToLive ?? s_defaultTimeToLive;
+
+            _cachedSasToken = null;
+        }
+
+        public string GetSasToken()
+        {
```

Comment:
Can a user provide sastoken?

Reply:
In the current implementation, no. User only provides a connection string, and the sdk evaluates the sas token based on that. The Azure SDK guidelines say: YOU MAY offer a way to create credentials from a connection string only if the service offers a connection string via the Azure portal.

Reply:
What does the Azure SDK guidelines say about users wanting to provide sastoken?

Reply:
I didn't find any specific reference to sas tokens, they say: Typically, binding parameters would include a URI to the service endpoint and authorization credentials.
This is something we are hoping to get some clarity on from Krzyszstof (I'll reach out to him after our meeting with Andrei on Monday).

Reply:
Tracked here: https://msazure.visualstudio.com/One/_workitems/edit/7456087